cmake/AutodetectLibrary.cmake

```diff
@@ -0,0 +1,164 @@
+
+                list(APPEND PKGCONFIG_LIBRARIES ${FIND_PKGCONFIG_LIBRARY})
+            endforeach()
+            unset(FIND_PKGCONFIG_LIBRARY CACHE)
+
+            set(${NAME}_LIBRARIES "${PKGCONFIG_LIBRARIES}")
+        endif (${NAME}_FOUND)
+    endif (PKG_CONFIG_FOUND)
+endmacro()
+
+macro(_autodetect_library_via_find NAME HEADER_PATH HEADER LIBRARY)
+    find_path(${NAME}_INCLUDE_DIR NAMES "${HEADER}" PATH_SUFFIXES "${HEADER_PATH}")
+    find_library(${NAME}_LIBRARY NAMES "${LIBRARY}")
+
+    if (${NAME}_INCLUDE_DIR AND ${NAME}_LIBRARY)
+        set(${NAME}_FOUND YES)
+        set(${NAME}_INCLUDE_DIRS "${${NAME}_INCLUDE_DIR}")
+        set(${NAME}_INCLUDE_LIBRARIES "${${NAME}_LIBRARY}")
+    endif (${NAME}_INCLUDE_DIR AND ${NAME}_LIBRARY)
+endmacro()
+
+macro(_patch_older_cmake NAME)
+    # Patch up libraries that still announce under the deprecated name only
+    if (NOT ${NAME}_LIBRARIES)
+        set(${NAME}_LIBRARIES "${${NAME}_LIBRARY}")
+    endif (NOT ${NAME}_LIBRARIES)
+    if (NOT ${NAME}_INCLUDE_DIRS)
+        set(${NAME}_INCLUDE_DIRS "${${NAME}_INCLUDE_DIR}")
+    endif (NOT ${NAME}_INCLUDE_DIRS)
+endmacro()
+
+macro(_patch_vcpkg_debug_optimized NAME)
+    # With vcpkg, the library path should contain both 'debug' and 'optimized'
+    # entries (see target_link_libraries() documentation for more information)
+    #
+    # Some libraries do this, but not all. In cases where this goes wrong,
+    # we will fix things up ourself.
+    #
+    # NOTE: we only patch up when using vcpkg; the same issue might happen
+    # when not using vcpkg, but this is non-trivial to fix, as we have no idea
+    # what the paths are. With vcpkg we do. And we only official support vcpkg
+    # with Windows.
+    #
+    # NOTE: this is based on the assumption that the debug file has the same
+    # name as the optimized file. This is not always the case, but so far
+    # experiences has shown that in those case vcpkg CMake files do the right
+    # thing.
+
+    if (VCPKG_TOOLCHAIN AND ${NAME}_LIBRARIES)
+        # Is this entry already of the correct format?
+        string(FIND "${${NAME}_LIBRARIES}" "optimized;" HAS_OPTIMIZED)
+        unset(PATCH_LIBRARIES)
+
+        if (HAS_OPTIMIZED LESS 0)
+            foreach(PATCH_LIBRARY ${${NAME}_LIBRARIES})
+                # Check if we have a 'debug' entry or an 'optimized' entry, and
+                # create the complementary entry.
+                if (${PATCH_LIBRARY} MATCHES "/debug/")
+                    set(PATCH_LIBRARY_DEBUG "${PATCH_LIBRARY}")
+                    string(REPLACE "/debug/lib/" "/lib/" PATCH_LIBRARY_RELEASE "${PATCH_LIBRARY}")
+                else (${LIBRARY} MATCHES "/debug/")
+                    set(PATCH_LIBRARY_RELEASE "${PATCH_LIBRARY}")
+                    string(REPLACE "/lib/" "/debug/lib/" PATCH_LIBRARY_DEBUG "${PATCH_LIBRARY}")
+                endif (${PATCH_LIBRARY} MATCHES "/debug/")
+
+                LIST(APPEND PATCH_LIBRARIES
```

Comment:
cmake --help-module SelectLibraryConfigurations

Reply:
The issue here really seems to be buggy stuff in `vcpkg`. Sometimes we see libraries that only return `debug` libraries, and other times only `optimized`. Not sure how `select_library_configurations()` is going to help here? (I am happy to be shown wrong btw; this is really annoying code to have, and to maintain).

Reply:
it can help setting the complete library variable from *_RELEASE and *_DEBUG variables.

Reply:
And the issue is only when generating a MSVC project (multi configuration), single configuration using Ninja generator works as intended.

Reply:
The problem exists only for MSVC as it has distinct C++ runtimes for release and debug.

Reply:
Ninja builds using MSVC compiler too, but vcpkg returns the correct debug/release lib in this case as only one kind is searched. It's really only when both kind are wanted that vcpkg fails, and not for all libs.